ARCs/arc-0065.md

```diff
@@ -0,0 +1,169 @@
+
+It is **RECOMMENDED** to use `UTF-8` for the error bytes string encoding.
+
+It is **RECOMMENDED** to use _short_ error messages.
+
+It is **RECOMMENDED** to use <a href="https://en.wikipedia.org/wiki/Camel_case/">camel case</a>
+for alphanumeric error codes.
+
+It is **RECOMMENDED** to avoid error byte strings of _exactly_ 8 or 32 bytes.
+
+### In Program Errors
+
+When a program wants to emit informative run time errors, directly from the bytecode,
+it **MUST**:
+
+1. Push to the stack the bytes string containing the error;
+1. Execute the `log` opcode to use the bytes from the top of the stack;
+1. Execute the `err` opcode to immediately terminate the program.
+
+Upon a program run time failure, the Algod API response contains both the failed
+_program counter_ (`pc`) and the `logs` array with the _errors_.
+
+The program **MAY** return multiple errors in the same failed execution.
+
+The errors **MUST** be retrieved by:
+
+1. Decoding the `base64` elements of the `logs` array;
+1. Validating the decoded elements against the error regexp.
+
+### Error examples
+
+> Error conforming this specification are always prefixed with `ERR:`.
+
+Error with a _numeric code_: `ERR:042`.
+
+Error with an _alphanumeric code_: `ERR:BadRequest`.
+
+Error with a _numeric code_ and _short message_: `ERR:042:AFunnyError`.
+
+### Program example
+
+The following program example raises the error `ERR:001:Invalid Method` for any
+application call to methods different from `m1()void`.
+
+```teal
+#pragma version 10
+
+txn ApplicationID
+bz end
+
+method "m1()void"
+txn ApplicationArgs 0
+match method1
+byte "ERR:001:Invalid Method"
+log
+err
+
+method1:
+b end
+
+end:
+int 1
+```
+
+Full Algod API response of a failed execution:
+
+```json
+{
+    "data": {
+        "app-index":1004,
+        "eval-states": [
+            { 
+                "logs": ["RVJSOjAwMTpJbnZhbGlkIE1ldGhvZA=="]
+            }
+        ],
+        "group-index":0,
+        "pc":41
+    },
+    "message":"TransactionPool.Remember: transaction ESI4GHAZY46MCUCLPBSB5HBRZPGO6V7DDUM5XKMNVPIRJK6DDAGQ: logic eval error: err opcode executed. Details: app=1004, pc=41, opcodes=pushbytes 0x4552523a3030313a496e76616c6964204d6574686f64 // \"ERR:001:Invalid Method\"; log; err; label2:"
+}
+```
+
+The `logs` array contains the `base64` encoded error `ERR:001:Invalid Method`.
+
+The `logs` array **MAY** contain elements that are not errors (as specified by the
+regexp).
+
+It is **NOT RECOMMENDED** to use on the `message` field to retrieve errors.
+
+### AVM Compilers
+
+AVM compilers (and related tools) **SHOULD** provide two error compiling options:
+
+1. The one specified in this ARC as **default**;
+1. The one specified in [ARC-64](./arc-0064.md) as fallback, if compiled bytecode
+size exceeds the AVM limits.
+
+> Compilers **MAY** optimize for program bytecode size by storing the error prefixes
+in the `bytecblock` and concatenating the error message at the cost of some extra
+opcodes.
```

Comment:
Relying on `.logs` is great when you have them, but I believe a lot of SDKs would hide `.logs` when e.g. throwing from a sendRawTransaction. The error message is mostly guaranteed to be propagated through the SDK, backend->frontend in case of a backend executor, etc. Hiding the error code from the algod error message would significantly detract value imo.